File: JS/LinkedList/easy/p0206_ReverseLinkedList.js
// @ts-check

import { isLinkedListEqual, LinkedList, ListNode } from '../index.js';

// 題目鏈結
// https://leetcode.com/problems/reverse-linked-list/

// 題目說明
// Given the head of a singly linked list, reverse the list, and return the reversed list.
// Input: head = [1,2,3,4,5]
// Output: [5,4,3,2,1]

// 解題重點
// 需要額外的tmp保存curr.next，待curr反轉後，再透過tmp去移動到下一節點

// 解題思路
// T = 0
// null    1 -> 2 -> 3
//    ^pre ^head
//
// T = 1
// null    1 -> 2 -> 3
//    ^pre ^head/tmp
//
// null    1 -> 2 -> 3  (head = head.next)
//    ^pre ^tmp ^head
//
// null <- 1    2 -> 3 (tmp.next = pre)
//    ^pre ^tmp ^head
//
// null <- 1    2 -> 3 (pre = tmp)
//         ^tmp ^head
//         ^pre
//
// T = 2
// null <- 1    2 -> 3
//         ^pre ^head/tmp
//
// null <- 1    2 -> 3
//         ^pre ^tmp ^head
//
// null <- 1 <- 2    3
//         ^pre ^tmp ^head
//
// null <- 1 <- 2    3
//              ^tmp ^head
//              ^pre
//
// T = 3
// null <- 1 <- 2    3 -> null
//              ^pre ^head/tmp
//
// null <- 1 <- 2    3 -> null
//              ^pre ^tmp ^head
//
// null <- 1 <- 2 <- 3    null
//              ^pre ^tmp ^head

// null <- 1 <- 2 <- 3    null
//                   ^tmp ^head
//                   ^pre
//
// 最後，head 指向null,  tmp, pre 皆可作為反轉後串列的新頭節點； 我們返回 pre 作為答案。

// 複雜度
// Time Complexity : O(N)
// Space Complexity: O(1)

/**
 * Definition for singly-linked list.
 * */
// function ListNode(val, next) {
//   this.val = val === undefined ? 0 : val;
//   this.next = next === undefined ? null : next;
// }

/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function (head) {
  let pre = null;  //反轉後 pre 會成為新的head後返回
  while (head) {
    let tmp = head;  // tmp 就像一個錨節點
    head = tmp.next; // tmp 幫助 head 移動的下一節點
    tmp.next = pre;  // tmp 反轉指針指向 pre
    pre = tmp;       // tmp 幫助 pre 移動到下一節點，完成一次循環。
  }
  return pre;
};

// 測試
(function () {
  console.log('Testing p0206_ReverseList...');

  console.log(
    isLinkedListEqual(
      reverseList(LinkedList([1, 2, 3, 4, 5])),
      LinkedList([5, 4, 3, 2, 1])
    ) === true
  );

  console.log('All Testing Passed ✅');
})();

File: Rust/src/problems/p0206_reverse_linked_list.rs
// 解題思路
//
// let mut head = Some(Box::new(Node {
//     val: 1,
//     next: Some(Box::new(Node {
//         val: 2,
//         next: Some(Box::new(Node {
//             val: 3,
//             next: Some(Box::new(Node {
//                 val: 4,
//                 next: Some(Box::new(Node { val: 5, next: None })),
//             })),
//         })),
//     })),
// }));
//
// T = 0
//
// res        head
// [{None}]   [ {1, next->[..]} ]
//
//
// T = 1
//
// res                       head                    node
// [{None}]                  [               ]       {1, next->[2..]}
//
//                                                   // 透過 let Some(node) = head，取出node
//
// res                       head                    node
// [{None}]                  [{2, next->[3..]}]      {1, next->[  ]}
//
//                                                   //head = node.next;
//
// res                       head                    node
// [      ]                  [{2, next->[3..]}]      {1, next->[{None}]}
//
//                                                   // node.next = res;
//
// res                       head                    node
// [{1, next->[{None}]}]     [{2, next->[3..]}]      None
//


// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

pub struct Solution {}
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut res = None;
        let mut head = head;
        while let Some(mut node) = head {
            head = node.next;
            node.next = res;
            res = Some(node);
        }
        res
    }
}

#[allow(dead_code)]
pub fn main() {
    fn get(v: Vec<i32>) -> Option<Box<ListNode>> {
        let mut head: Option<Box<ListNode>> = None;
        let mut curr = &mut head;

        for e in v {
            curr = &mut curr.insert(Box::new(ListNode::new(e))).next;
        }

        head
    }

    fn is_equal(mut l1: Option<Box<ListNode>>, mut l2: Option<Box<ListNode>>) -> bool {
        while l1.is_some() && l2.is_some() {
            if l1.as_ref().unwrap().val != l2.as_ref().unwrap().val {
                return false;
            }
            l1 = l1.unwrap().next;
            l2 = l2.unwrap().next;
        }

        if l1.is_some() || l2.is_some() {
            return false;
        }

        return true;
    }
    fn test_case(v1: Vec<i32>, expect: Vec<i32>) -> bool {
        let ans = get(expect);
        let out = Solution::reverse_list(get(v1));
        is_equal(out, ans)
    }

    println!("Start p0206_reverse_list testing...");

    println!("{}", test_case(vec![1, 2, 3, 4, 5], vec![5, 4, 3, 2, 1]));
    println!("{}", test_case(vec![1, 2], vec![2, 1]));
    println!("{}", test_case(vec![], vec![]));

    println!("All Testing Passed ✅");
}


